### PR TITLE
Esp32c6 Vial and storage works

### PR DESCRIPTION
**Pull request by davidqge**
Last change Also improved full sized keyboard key loading time.

### CODE REVIEW

**Comment by HaoboGu** [comment]
Thank you a lot! 

I just moved `src/storage.rs` to `src/storage/mod.rs`, could you merge latest main to your branch?

**Comment by davidqge** [comment]
> I just moved `src/storage.rs` to `src/storage/mod.rs`, could you merge latest main to your branch?

done

File: rmk/Cargo.toml
Comment (HaoboGu) [inline rmk/Cargo.toml]: I can merge this now, but using a git dependency will block the release to crates.io. So it's better to make it released to crates.io soon then we can switch to crates.io version.

File: rmk/Cargo.toml
Comment (HaoboGu) [inline rmk/Cargo.toml]: I prefer to remove this dep, keep the dependencies tree clean and use plain `as` for converting primitives.

File: rmk/Cargo.toml
Comment (HaoboGu) [inline rmk/Cargo.toml]: Same with `saturating_cast`, I prefer to remove this and use plain implementation. That will make code more direct and readable.

File: rmk/src/hid.rs
Comment (HaoboGu) [inline rmk/src/hid.rs]: Make tupe `(T1, T2)` as a separate struct?

File: rmk/src/hid.rs
Comment (HaoboGu) [inline rmk/src/hid.rs:75]: Set `ConnectionType::Ble` to embassy's `channel:Receiver` is quite misleading. As I mentioned above, it's better to make a new struct and impl these traits for that struct type.

**Comment by HaoboGu** [comment]
This PR looks good overall, I leave some reviews mainly for maintainability and modularity. Feel free to comment if you have any questions!

File: rmk/Cargo.toml
Comment (davidqge) [reply rmk/Cargo.toml]: I removed last commit, so the other commits can be merged now.

Comment (HaoboGu) [reply rmk/Cargo.toml]: cool, I can merge this if other  comments are resolved

**Comment by HaoboGu** [comment]
btw I'm curious that does vial support BLE? I thought it doesn't, so I use vial only via USB. If vial supports BLE, I'd like to make all RMK devices use vial over BLE

**Review by HaoboGu — CHANGES_REQUESTED** [review]

**Comment by davidqge** [comment]
> I'm thinking how to make vial-web recognize RMK BLE keyboard as well, then we can deploy one, which could be quite useful.

Vial web can find the device, but their desktop version send message with report id 0, even though descriptor says 5. I can't get that message from esp32-nimble. So I have to change Vial desktop to send on report id 5.

Let me know if other device allows you to get the report id 0.

**Comment by HaoboGu** [comment]
> Vial web can find the device, but send message with report id 0, even though descriptor says 5

ReportID on vial-web is hard-coded to 0:

https://github.com/vial-kb/vial-web/blob/b8b8f97e029b349e7137e2a95a166449814f9ee3/src/index.html#L107-L109

File: rmk/Cargo.toml
Comment (davidqge) [reply rmk/Cargo.toml]: Changed as requested.
as don't do saturating cast for integer. 
 it makes overflow management a little easier, and code more robust.

File: rmk/src/hid.rs
Comment (HaoboGu) [inline rmk/src/hid.rs:47]: What about doing:

```rust
pub(crate) struct EspBleReaderWriter<'ch, M: RawMutex, T: Sized, const N: usize, W: HidWriterWrapper> {
    receiver: Receiver<'ch, M, T, N>,
    hid_writer: W,
}

impl<'ch, M: RawMutex, T: Sized, const N: usize, W: HidWriterWrapper> ConnectionTypeWrapper
    for EspBleReaderWriter<'ch, M, T, N, W>
{
    fn get_conn_type(&self) -> ConnectionType {
        ConnectionType::Ble
    }
}

impl<'ch, M: RawMutex, T: Sized, const N: usize, W: HidWriterWrapper> HidReaderWrapper
    for EspBleReaderWriter<'ch, M, T, N, W>
{
    async fn read(&mut self, buf: &mut [u8]) -> Result<usize, HidError> {
        let v = self.receiver.receive().await;
        buf.copy_from_slice(as_bytes(&v));
        Ok(as_bytes(&v).len())
    }
}

impl<'ch, M: RawMutex, T: Sized, const N: usize, W: HidWriterWrapper> HidWriterWrapper
    for EspBleReaderWriter<'ch, M, T, N, W>
{
    async fn write_serialize<IR: AsInputReport>(&mut self, r: &IR) -> Result<(), HidError> {
        self.hid_writer.write_serialize(r).await
    }

    async fn write(&mut self, report: &[u8]) -> Result<(), HidError> {
        self.hid_writer.write(report).await
    }
}
```
And put it to `esp` mod?

Then you can use it as:

```rust
let mut via_rw = EspBleReaderWriter::new(via_output.receiver(), ble_server.input_vial);
```

What do you think?

**Review by HaoboGu — APPROVED** [review]

**Pull request merged**
